### ISSUE TITLE
ray.init() error

### ISSUE THREAD

**Issue by ArijitSinghEDA** [ray-project/ray#43380]
### What happened + What you expected to happen

I ran simple two lines
```
import ray
ray.init()
```

I get this error:
```
2024-02-22 13:57:06,461 WARNING worker.py:1729 -- Failed to set SIGTERM handler, processes mightnot be cleaned up properly on exit.
2024-02-22 13:57:32,647 ERROR services.py:1207 -- Failed to start the dashboard
2024-02-22 13:57:32,648 ERROR services.py:1232 -- Error should be written to 'dashboard.log' or 'dashboard.err'. We are printing the last 20 lines for you. See 'https://docs.ray.io/en/master/ray-observability/ray-logging.html#logging-directory-structure' to find where the log file is.
2024-02-22 13:57:32,648 ERROR services.py:1242 -- Couldn't read dashboard.log file. Error: [Errno 2] No such file or directory: '/tmp/ray/session_2024-02-22_13-57-06_577198_21267/logs/dashboard.log'. It means the dashboard is broken even before it initializes the logger (mostly dependency issues). Reading the dashboard.err file which contains stdout/stderr.
2024-02-22 13:57:32,648 ERROR services.py:1276 -- Failed to read dashboard.err file: cannot mmap an empty file. It is unexpected. Please report an issue to Ray github. https://github.com/ray-project/ray/issues
[2024-02-22 13:58:03,776 E 21267 21372] core_worker.cc:201: Failed to register worker 01000000ffffffffffffffffffffffffffffffffffffffffffffffff to Raylet. 
IOError: [RayletClient] Unable to register worker with raylet. No such file or directory
```

It should have started a ray instance.

### Versions / Dependencies

Python: 3.10.12
Ray: 2.6.3
grpcio: 1.58.0
WSL: Ubuntu 22.04.2 LTS (Jammy Jellyfish)

### Reproduction script

```
import ray
ray.init()
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by anyscalesam** [comment]
@ArijitSinghEDA can you attempt repro on latest ray293?

**Comment by ArijitSinghEDA** [comment]
@anyscalesam I need to run it on ray2.6.3 due to some limitations, so version upgrade is not the solution for me

**Comment by Arijit-crypt** [comment]
@anyscalesam I believe it is an issue with WSL. When I ran the same code on pure Linux, it started without any error

**Comment by ArijitSinghEDA** [comment]
@anyscalesam I believe it is an issue with WSL. When I ran the same code on pure Linux, it started without any error

**Comment by anyscalesam** [comment]
@mattip please take a look and determine priority; reach out to core anyscaler team if you have questions

**Comment by mattip** [comment]
TL;DR: Maybe you are running WSL headless and missing allowing ray through the firewall?

And for the whole triage report:

I get a different error when I did the following on WSL Ubuntu 22.04. Note I got grpcio 1.62.1. I did get some firewall popup windows that required user intervention the first time I ran `ray.init()`, so the process cannot be run headless:
```
$ python3 -m venv /tmp/venv310
$ source /tmp/venv310/bin/activate

(venv310) $ pip install ray==2.6.3

(venv310) $ pip list | grep "\(grpcio\)\|\(ray\)"
grpcio                    1.62.1
ray                       2.6.3
 
(venv310) $ python
>>> import ray
>>> ray.init()
2024-03-11 09:10:47,742 INFO worker.py:1621 -- Started a local Ray instance.
E0311 09:10:48.908024500     306 socket_utils_common_posix.cc:223] check for SO_REUSEPORT: \
    {"created":"@1710141048.908010200","description":"Protocol not available",\
     "errno":92,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/socket_utils_common_posix.cc","file_line":202,\
     "os_error":"Protocol not available","syscall":"getsockopt(SO_REUSEPORT)"\
    }
RayContext(dashboard_url='', python_version='3.10.12', ray_version='2.6.3', ray_commit='8a434b4ee7cd48e60fa1531315d39901fac5d79e', protocol_version=None)
(raylet) E0311 09:10:48.881503200     410 socket_utils_common_posix.cc:223] check for SO_REUSEPORT:\
     {"created":"@1710141048.881474900","description":"Protocol not available",\
      "errno":92,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/socket_utils_common_posix.cc","file_line":202,\
      "os_error":"Protocol not available","syscall":"getsockopt(SO_REUSEPORT)"
```

So it seems `SO_REUSEPORT` is not available on WSL. If I force grpcio 1.58.0, I get the same result.

But this does not seem to stop ray from working:

<details><summary>test_ray.py</summary>

```
$ cat /tmp/test_ray.py
import ray
ray.init()


# Define the square task.
@ray.remote
def square(x):
    return x * x


# Launch four parallel square tasks.
futures = [square.remote(i) for i in range(4)]


# Retrieve results.
print(ray.get(futures))
```

</details>

```
$ python /tmp/test_ray.py
2024-03-11 09:37:12,120 INFO worker.py:1621 -- Started a local Ray instance.
E0311 09:37:13.347362400     942 socket_utils_common_posix.cc:223] check for SO_REUSEPORT: {"created":"@1710142633.347321900","description":"Protocol not available","errno":92,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/socket_utils_common_posix.cc","file_line":202,"os_error":"Protocol not available","syscall":"getsockopt(SO_REUSEPORT)"}
[0, 1, 4, 9]
```
. I do see an unguarded call in ray, maybe it should be conditional on `SO_REUSEPORT` working https://github.com/ray-project/ray/blob/4d2be8f35a805d251fa139838064a435e53fa2a2/src/ray/rpc/grpc_server.cc#L63-L67

I would think this use case is niche enough that fixing the error message is not a high priority.
